include/eve/algo/container/detail/soa_storage2.hpp

```diff
@@ -0,0 +1,189 @@
+    {
+      auto ptrs = kumi::map([]( auto p) { return as_aligned_pointer( p ); }, *this);
+      return kumi::apply(eve::algo::views::zip, ptrs);
+    }
+
+    EVE_FORCEINLINE auto data()       { return eve::unalign(data_aligned()); }
+    EVE_FORCEINLINE auto data() const { return eve::unalign(data_aligned()); }
+
+    void swap(soa_storage& other) noexcept
+    {
+      storage_.swap(other.storage_);
+      std::swap(indexes_  , other.indexes_  );
+      std::swap(capacity_ , other.capacity_ );
+      std::swap(size_     , other.size_     );
+    }
+
+    friend void swap(soa_storage& a, soa_storage& b ) noexcept { a.swap(b); }
+
+    //==============================================================================================
+    // Static helpers
+    //==============================================================================================
+    template<typename T> static EVE_FORCEINLINE auto as_aligned_pointer(T* ptr)
+    {
+      return eve::as_aligned(ptr, eve::detail::cache_line_cardinal<T>{});
+    }
+
+    template<typename T>
+    static  auto realign(std::size_t n) { return eve::align(sizeof(T)*n, eve::over{64}); }
+
+    static auto allocate(Allocator const& a,std::size_t n)
+    {
+      aligned_deleter d{a};
+      return storage_t{ reinterpret_cast<byte_t*>(d.allocate_aligned(n,64)), d};
+    }
+
+    static auto aligned_capacity(std::size_t n) noexcept
+    {
+      // Capacity is the smallest gap possible, ie the one in the largest type
+      constexpr std::size_t largest = max_scalar_size_v<Type>;
+      return eve::align(largest*n, eve::over{64})/largest;
+    }
+
+    static auto byte_size(std::size_t n) noexcept
+    {
+      return  kumi::fold_left ( [n]<typename S>(auto a, S) { return a + realign<S>(n); }
+                              , flat_t{}
+                              , 0ULL
+                              );
+    }
+
+    //==============================================================================================
+    // Tuple interface
+    //==============================================================================================
+    template<std::size_t I>
+    friend decltype(auto) get(soa_storage& s) noexcept
+    {
+      return reinterpret_cast<kumi::element_t<I,Type>*>(s.storage_.get() + get<I>(s.indexes_));
+    }
+
+    template<std::size_t I>
+    friend decltype(auto) get(soa_storage const& s) noexcept
+    {
+      return reinterpret_cast<kumi::element_t<I,Type> const*>(s.storage_.get() + get<I>(s.indexes_));
+    }
+
+    //==============================================================================================
+    template<typename U> struct as_index { using type = std::size_t; };
+
+    struct aligned_deleter : Allocator
+    {
+      aligned_deleter(Allocator const& a) : Allocator(a) {}
+      void operator()(void* p) { if(p) Allocator::deallocate_aligned(p); }
+    };
+
+    Allocator& get_allocator() { return storage_.get_deleter(); }
```

Comment:
const. And by value, allocators are passed by value.

Reply:
Fixed